### ISSUE TITLE
Why is the loss for the trajectories in each episode not decreasing for DQN example? 

### ISSUE THREAD

**Issue by alejandroarmas** [deepmind/acme#51]
Hello everyone, I ran the `acme/run_dqn.py` example script and I have noticed that while the episode return increases but the learner loss is not decreasing as the episodes continue? Why is that? What logs are at my disposal in order to see if agents are improving in future episodes? 

Here is the output after a few episodes as opposed to 895 episodes in. 
```console
[Environment Loop] Episode Length = 764 | Episode Return = -21.000 | Episodes = 2 | Steps = 1528 | Steps Per Second = 37.277
[Learner] Loss = 0.058 | Steps = 70 | Walltime = 4.277
[Learner] Loss = 0.058 | Steps = 87 | Walltime = 5.307
[Learner] Loss = 0.163 | Steps = 104 | Walltime = 6.357
[Learner] Loss = 0.073 | Steps = 121 | Walltime = 7.372
[Learner] Loss = 0.068 | Steps = 138 | Walltime = 8.380
[Learner] Loss = 0.061 | Steps = 155 | Walltime = 9.391
[Learner] Loss = 0.052 | Steps = 173 | Walltime = 10.452
[Environment Loop] Episode Length = 876 | Episode Return = -21.000 | Episodes = 3 | Steps = 2404 | Steps Per Second = 133.777
[Learner] Loss = 0.054 | Steps = 190 | Walltime = 11.485
[Learner] Loss = 0.060 | Steps = 207 | Walltime = 12.543
```
Very quickly the loss decreases...
```console
[Environment Loop] Episode Length = 764 | Episode Return = -21.000 | Episodes = 6 | Steps = 4696 | Steps Per Second = 140.916
[Learner] Loss = 0.009 | Steps = 467 | Walltime = 27.883
[Learner] Loss = 0.014 | Steps = 485 | Walltime = 28.910
[Learner] Loss = 0.014 | Steps = 503 | Walltime = 29.950
[Learner] Loss = 0.013 | Steps = 520 | Walltime = 30.955
[Learner] Loss = 0.020 | Steps = 537 | Walltime = 31.968
[Learner] Loss = 0.016 | Steps = 554 | Walltime = 33.019
[Learner] Loss = 0.011 | Steps = 571 | Walltime = 34.030
[Environment Loop] Episode Length = 912 | Episode Return = -21.000 | Episodes = 7 | Steps = 5608 | Steps Per Second = 135.071
[Learner] Loss = 0.010 | Steps = 588 | Walltime = 35.087
[Learner] Loss = 0.014 | Steps = 604 | Walltime = 36.109
[Learner] Loss = 0.015 | Steps = 615 | Walltime = 37.127
[Learner] Loss = 0.016 | Steps = 629 | Walltime = 38.193

```

But seems to not improve much for the rest of the training loop
```console
[Environment Loop] Episode Length = 2154 | Episode Return = 12.000 | Episodes = 895 | Steps = 1838193 | Steps Per Second = 6.382
[Learner] Loss = 0.011 | Steps = 229651 | Walltime = 304473.760
[Learner] Loss = 0.009 | Steps = 229652 | Walltime = 304475.012
[Learner] Loss = 0.012 | Steps = 229653 | Walltime = 304476.288
[Learner] Loss = 0.008 | Steps = 229654 | Walltime = 304477.543
[Learner] Loss = 0.009 | Steps = 229655 | Walltime = 304478.783
[Learner] Loss = 0.008 | Steps = 229656 | Walltime = 304480.020
[Learner] Loss = 0.010 | Steps = 229657 | Walltime = 304481.291
[Learner] Loss = 0.017 | Steps = 229658 | Walltime = 304482.534
[Learner] Loss = 0.009 | Steps = 229659 | Walltime = 304483.770

```

**Comment by fmxFranky** [comment]
I think maybe you should do at least the following two things and try again if you only run the current py file provided by acme repo. (I'm not a contributor, and i don't make sure whether they are usefull to reimplement DQN's metrics in the paper)
- Rewrite the code in `Actor` class, because the code in `acme.agent.DQN` don't give the true epsilon-greedy policy, which only provides a constant epsilon in `DQN` class. In another issue the contibutor also said this.
- Change the hyperparameters to fit the numbers provided in the DQN paper

**Comment by mokeddembillel** [comment]
I'm not 100% sure, but i think the reason why the loss is not decreasing, is that in reinforcement learning with function approximation (DQN in your case), is not the same as in supervised learning where the dataset is fixed and known from  the beginning, in RL case the data is being generated by the agent and gathered from the replay buffer, so this problem may occur because of the neural network learning from some experiences more than others, in more details read this scenario:
- at first the replay  buffer is going to be incomplete so sampling experiences multiple times from it will make your neural network learn more from a small set of experiences than others, i think this may result in an overfitting (which means a huge decrease in learning loss) and after the replay buffer becomes full the neural network start learning from new and diverse experiences, so the loss start increasing again, i did some tests (check my repo: https://github.com/mokeddembillel/Lunar-Lander-Deep-Expected-Sarsa) and i found out that the variations of the loss could be controlled by these parameters:
     - the replay buffer size
     - the number of replay steps per agent step
     - the mini-batch size (to clarify i'm talking about the mini-batch that we sample from  the replay buffer each replay step
     - the step size (the learning rate)

and to know if the agent is progressing or not, i would recommend that you focus more on the reward than the loss and in some cases the number of steps taken by the agent to finish an episode is worth considering if you want your agent to finish the episode as fast as possible 

**Please correct me if i'm wrong at something** :) and **Good Luck**

**Comment by nikolamomchev** [comment]
This issue is quite stale and I believe it was probably answered sufficiently well, so I am closing it.

If you still have questions, feel free to submit them and we'll try to answer to the best of our ability.
